Documentation (#25)
vimeo/dials: README.md

only the value from command line flag populates the config due to the special `dialsflag` tag. If you prefer the value for `Val3` be overwritten by other sources, then
use the `dials` tag instead of the `dialsflag` tag

### Watching file source
If you wish to watch the config file and make updates to your configuration, use the watching source. This functionality is available in the `ez` package by passing the `true` boolean to the functions. The `WatchingSource` can be used when you want to further customize the configuration as well.

``` go
	// NewWatchSource also has watch options that can be passed to have the
	// ability to use a ticker for polling, set a logger, and more
	watchingFileSource, fsErr := file.NewWatchingSource(
		"path/to/config", &yaml.Decoder{})

	if fsErr != nil {
		// error handling
		return
	}

	// additional sources can be passed along with the watching file source and the
	// precedence order will still be dictated by the order in which the sources are
	// defined in the Config function.
	view, err := dials.Config(context.Background(), config, watchingFileSource)
	if err != nil {
		// error handling
	}

	Config := view.View().(*Config)
```

### Source

> Source interface is implemented by different configuration sources that populate the configuration struct. Dials currently supports environment variables, command line flags, and config file sources. When `dials.Config` function is going through the different sources to extract the values, it calls the `Value` method on each of these sources. This allows for the logic of the source to be encapsulated while giving the application access to the values populated by each source. The `dials.Config` function composes the final config struct by overlaying the values from all the different sources and accounting for the precedence order. If you wish to define your own source, implement the `Source` interface and pass the source to the `dials.Config` function.

Reviewer: We should probably break this section up into a few paragraphs. (and probably have a subsection about writing a `Source`). -- including a mention of when it's appropriate to delegate to a `Decoder` implementation.

We should probably also include a mention of pointerification in a "nitty-gritty" section or something, so it's clear how we actually implement overlaying.

### Decoder

> Decoder interface is implemented by different data formats to decode the data and insert the values into the appropriate fields in the config struct. Dials currently supports JSON, YAML, and TOML data formats. If you wish to decode another data format, implement the `Decoder` interface and pass the decoder to the sources that support decoding (string and file sources). When the `dials.Config` function calls the `Value` method for the source, the supported source will call the `Decode` method to unmarshal the data into the config struct and return the populated struct.

Reviewer: I feel like this paragraph is missing an explanation of why `Decoder` exists.

// Dials will first read from environment variables and command line flags
// and then read the config file specified by the populated field.
func (c *Config) ConfigPath() (string, bool) {
	// can alternatively return empty string and false if the state of the
	// struct doesn't specify a config file to read
	return c.Path, true
}

func main() {
	c := &Config{}

	// The following function will populate the config struct by reading the
	// config files, environment variables, and command line flags (order matches
	// the function name) with increasing precedence. In other words, the flag
	// source (last) would overwrite the YAML source (first) were they both to
	// attempt to set the same struct field. The boolean argument passed to the
	// function indicates whether the file will be watched and updates to the
	// file should update the config struct.
	view, dialsErr := ez.YAMLConfigEnvFlag(context.Background(), c, false)
	if dialsErr != nil {
		// error handling
	}

	// Get an interface corresponding to the filled-out config struct, and
	// assert it to the correct type. Here's the struct populated from config file,
	// environment variables, and command line flags.

> 	Config := view.View().(*Config)

Reviewer 2: A bit strange to have a variable with an initial capital letter; maybe just call this `config`.

Also, it might be nice to show an example of the `Fill` methods for those that might be put off by the necessary type assert (until we get Generics!):
```go
config := &Config{}
d.Fill(config)
```

only the value from command line flag populates the config due to the special `dialsflag` tag. If you prefer the value for `Val3` be overwritten by other sources, then
use the `dials` tag instead of the `dialsflag` tag

### Watching file source
If you wish to watch the config file and make updates to your configuration, use the watching source. This functionality is available in the `ez` package by passing the `true` boolean to the functions. The `WatchingSource` can be used when you want to further customize the configuration as well.

``` go
	// NewWatchSource also has watch options that can be passed to have the
	// ability to use a ticker for polling, set a logger, and more
	watchingFileSource, fsErr := file.NewWatchingSource(
		"path/to/config", &yaml.Decoder{})

	if fsErr != nil {
		// error handling
		return
	}

	// additional sources can be passed along with the watching file source and the
	// precedence order will still be dictated by the order in which the sources are
	// defined in the Config function.
	view, err := dials.Config(context.Background(), config, watchingFileSource)
	if err != nil {
		// error handling
	}

	Config := view.View().(*Config)
```

### Source

> Source interface is implemented by different configuration sources that populate the configuration struct. Dials currently supports environment variables, command line flags, and config file sources. When `dials.Config` function is going through the different sources to extract the values, it calls the `Value` method on each of these sources. This allows for the logic of the Source to be encapsulated while giving the application access to the values populated by each Source.

Reviewer 2: I think *The* Source interface is implemented... sounds more natural to me.

# Dials

[![Actions Status](https://github.com/vimeo/dials/workflows/Go/badge.svg)](https://github.com/vimeo/dials/actions)
[![GoDoc](https://godoc.org/github.com/vimeo/dials?status.svg)](https://godoc.org/github.com/vimeo/dials)
[![Go Report Card](https://goreportcard.com/badge/github.com/vimeo/dials)](https://goreportcard.com/report/github.com/vimeo/dials)

Dials is an extensible configuration package for Go.

## Installation

```
go get github.com/vimeo/dials
```

## Prerequisites

Dials requires Go 1.13 or later.

> ## What is Dials

Reviewer: Perhaps this is a question?

Just populate the config struct with the default values and pass the struct to Dials. 
Dials also allows the flexibility to choose the precedence order to determine which sources can overwrite the configuration values. Additionally, Dials has special handling of structs that implement [`encoding.TextUnmarshaler`](https://golang.org/pkg/encoding/#TextUnmarshaler) so structs (like [`IP`](https://pkg.go.dev/net?tab=doc#IP) and [`time`](https://pkg.go.dev/time?tab=doc#Time)) can be properly parsed.

## Using Dials

### Reading from config files, environment variables, and command line flags
Dials requires very minimal configuration to populate values from several sources through the `ez` package. Define the config struct and provide a method `ConfigPath() (string, bool)` to indicate the path to the config file. The package defines several functions that help populate the config struct by reading from config files,
environment variables, and command line flags

``` go
package main

import (
	"context"
	"fmt"

	"github.com/vimeo/dials/ez"
)

// Config is the configuration struct needed for the application

> type Config struct {

Reviewer: let's put `dialsdesc` tags on at least one of the fields.

	// When a struct tag more specifically corresponding to a source is
	// present, it takes precedence over the `dials` tag. Note that just
	// because there is a `yaml` struct tag present doesn't mean that other
	// sources can't fill this field.
	Val1 string `dials:"Val1" yaml:"b"`
	// the dials tag can be used as an alias so when the name in the config file
	// changes, the code doesn't have to change.
	Val2 int `dials:"val_2"`
	// Dials follows the Go convention for flags and will look for the dials
	// tag or field name in lower-kebab-case. Without any struct tags, it

> 	// would look for val-3 flag. To specify a different flag name, use the

Reviewer: Hmm, Go doesn't have a particularly strong flag-naming convention. Either kebab or snake case are common.

Also, I wouldn't describe it as dial will "look for" a flag, so much as, it will "register a flag" with that name. (as you know, the flag and pflag sources iterate over set flags)

	// `dialsflag` tag. Now, Dials will lookup "some-val" flag instead.
	// Since only `dialsflag` tag is specified, Val3 will only be populated
	// from command line flags
	Val3 bool `dialsflag:"some-val"`
	// Path holds the value of the path to the config file. Dials follows the
	// Go convention and will look for the dials tag or field name in all caps
	// when struct tags aren't specified. Without any struct tags, it would
	// lookup the PATH environment variable. To specify a different env variable,
	// use the `dials_env` tag. Now Dials will lookup "configpath" env value to
	// populate the Path field
	Path string `dials_env:"configpath"`
}

> // ConfigPath returns the path to the config file. This is particularly helpful
> // when the path is populated from environment variables or command line flags.
> // Dials will first read from environment variables and command line flags
> // and then read the config file specified by the populated field.

Reviewer: I think this docstring could be slightly clearer:

```suggestion
// ConfigPath returns the path to the config file that dials should read. This is particularly helpful
// when it's desirable to specify the file's path via environment variables or command line flags.
// Dials will first populate the configuration struct from environment variables and command line flags
// and then read the config file that the `ConfigPath()` method returns.
```

func (c *Config) ConfigPath() (string, bool) {
	// can alternatively return empty string and false if the state of the

> 	// struct doesn't specify a config file to read

Reviewer: Hmm, users who just want to use environment variables and flags probably don't need the `ez` package's helpers. (it might actually be more confusing to use `ez` than not.)

Perhaps we should have a simple example of just the flag source and environment variable source being passed to dials. (definitely in another section, though.)

	return c.Path, true
}

func main() {
	c := &Config{}

	// The following function will populate the config struct by reading the
	// config files, environment variables, and command line flags (order matches
	// the function name) with increasing precedence. In other words, the flag
	// source (last) would overwrite the YAML source (first) were they both to
	// attempt to set the same struct field. There are several options that can be
	// passed in as well to indicate whether the file will be watched and updates
	// to the file should update the config struct and if the flags name component
	// separation should use different encoding

> 	d, dialsErr := ez.YAMLConfigEnvFlag(context.Background(), c, WithWatchingConfigFile())

Reviewer: nit: with Sergio's PR, we need an argument to `WithWatchingConfigFile`. Also, we need a package reference on that function call.

> 	d, dialsErr := ez.YAMLConfigEnvFlag(context.Background(), c, WithWatchingConfigFile())

Reviewer: Hmm, one thing I realized when writing a test for `dials.Config` the other day is that examples involving a watching-source should always provide contexts derived from  `context.WithCancel()` so the watching goroutine doesn't leak.

Hmm, that means that we're leaking a goroutine in the `ez` package. (I'll put together a PR to fix that after Sergio's is PR #32 is merged)

``` yaml
b: valueb
val_2: 2
val-3: false
```

and the following command (make sure to change the configpath value to point to your path)

```
export configpath=path/to/config/file
export VAL_2=5
go run main.go --some-val
```

the output will be 
`Config: &{Val1:valueb Val2:5 Val3:true}`

Note that even though val_2 has a value of 2 in the yaml file, the config value 
output for Val2 is 5 because environment variables take precedence.

### Configure your configuration settings
If the predefined functions in the ez package don't meet your needs, you can specify the 
sources you would like in the order of precedence you prefer. Not much setup is needed 
to configure this. Choose the predefined sources and add the appropriate `dials` tags to the config struct. 

``` go
package main

import (
	"context"
	"fmt"

	"github.com/vimeo/dials"
	"github.com/vimeo/dials/env"
	"github.com/vimeo/dials/file"
	"github.com/vimeo/dials/flag"
	"github.com/vimeo/dials/yaml"
)

type Config struct {
	Val1 string `dials:"Val1" yaml:"b"`
	// The `dials_env` tag is used for environment values. The Val2 value will
	// only be populated from environment variables. If you want several different

> 	// sources to be able to set this value, use the `dials` tag instead

Reviewer: Hmm, this is misleading.
The `dials_env` and `dialsflag` tags don't prevent other sources from setting the field, they just override the name used by that source.

The second sentence should probably read something like: 
```
If you want to use a single, consistent name across several sources, set the `dials` tag instead
```

	Val2 int `dials_env:"VAL_2"`
	// the `dialsflag` tag is used for command line flag values. The Val3 value
	// will only be populated from command line flags

> 	Val3 bool `dialsflag:"val-3"`

Reviewer: Let's add an `dialsdesc` tag to this field to go with `dialsflag`.

}

func main() {
	config := &Config{
		// Val1 has a default value of "hello" and will be overwritten by the
		// sources if there is a corresponding field for Val1
		Val1: "hello",
	}

	// To read from other source files such as JSON, and TOML, use
	// "&json.Decoder{}" or "&toml.Decoder{}"
	fileSrc, fileErr := file.NewSource("path/to/config", &yaml.Decoder{})
	if fileErr != nil {
		// error handling
	}

	// Define a `dials.Source` for command line flags. Consider using the dials pflag library

> 	// if the application uses the spf13/pflag package

Reviewer: let's spell out the full, absolute package name for `pflag`

	}

	// use the environment source to get values from environment variables
	envSrc := &env.Source{}

	// the Config struct will be populated in the order in which the sources are
	// passed in the Config function with increasing precedence. So the fileSrc value
	// will overwrite the flagSet value if they both were to set for the same field
	d, err := dials.Config(context.Background(), config, envSrc, flagSet, fileSrc)
	if err != nil {
		// error handling
	}

	// Fill populates the config struct
	c := &Config{}
	d.Fill(c)
	fmt.Printf("Config: %+v\n", c)
}
```

If the above code is run with the following YAML file (make sure to change the path in the code):

``` yaml
b: valueb
val-3: false
```

and the following commands 

``` 
export VAL_2=5
go run main.go --val-3
```
 
the output will be `Config: &{Val1:valueb Val2:5 Val3:true}`. Note that even when val-3 is defined in the yaml file and the file source takes precedence, 
only the value from command line flag populates the config due to the special `dialsflag` tag. If you prefer the value for `Val3` be overwritten by other sources, then
use the `dials` tag instead of the `dialsflag` tag

### Watching file source

> If you wish to watch the config file and make updates to your configuration, use the watching source. This functionality is available in the `ez` package by passing the `true` boolean to the functions. The `WatchingSource` can be used when you want to further customize the configuration as well. Please note that the Watcher interface is likely to change in the near future.

Reviewer: Looks like this now needs to be updated with the new options args in `ez`

``` go

> 	// NewWatchSource also has watch options that can be passed to have the
> 	// ability to use a ticker for polling, set a logger, and more

Reviewer: let's drop `to have the ability to`, so this is:
```suggestion
	// NewWatchSource also has watch options that can be passed to
	// use a ticker for polling, set a logger, and more
```

	watchingFileSource, fsErr := file.NewWatchingSource(
		"path/to/config", &yaml.Decoder{})

	if fsErr != nil {
		// error handling
		return
	}

	// additional sources can be passed along with the watching file source and the
	// precedence order will still be dictated by the order in which the sources are
	// defined in the Config function.

> 	d, err := dials.Config(context.Background(), config, watchingFileSource)

Reviewer: Let's explicitly call out using a non-background context here and include a context coming from `context.WithCancel`.

	if err != nil {
		// error handling
	}

	d.Fill(config)

> ```

Reviewer: With the watching source, we should include waiting for the channel on the `Events()` method.

### Source

> The Source interface is implemented by different configuration sources that populate the configuration struct. Dials currently supports environment variables, command line flags, and config file sources. When `dials.Config` function is going through the different sources to extract the values, it calls the `Value` method on each of these sources. This allows for the logic of the Source to be encapsulated while giving the application access to the values populated by each Source. Please note that the Value method on the Source interface and the Watcher interface are likely to change in the near future.

Reviewer: nit:
```suggestion
The Source interface is implemented by different configuration sources that populate the configuration struct. Dials currently supports environment variables, command line flags, and config file sources. When the `dials.Config` method is going through the different `Source`s to extract the values, it calls the `Value` method on each of these sources. This allows for the logic of the Source to be encapsulated while giving the application access to the values populated by each Source. Please note that the Value method on the Source interface and the Watcher interface are likely to change in the near future.
```

### Decoder
Decoders are modular allowing users to mix and match Decoders and Sources. Dials currently supports Decoders that decode different data formats (JSON, YAML, and TOML) and insert the values into the appropriate fields in the config struct. Decoders can be expanded from that use case and users can write their own Decoders to perform the tasks they like (more info in the section below). Decoder is called when the supported Source calls the `Decode` method to unmarshal the data into the config struct and returns the populated struct. There are two sources that the Decoders can be used with: files (including watched files) and `static.StringSource`. Please note that the Decoder interface is likely to change in the near future.

### Write your own Source and Decoder

> If you wish to define your own source, implement the `Source` interface and pass the source to the `dials.Config` function. If you want the Source to interact with a Decoder, call `Decode` in the `Value` method of the Source. If you want to define your own Decoder to interact with your Source, implement the `Decoder` interface. Since Decoders are modular, keep the logic of Decoder encapsulated and separate from the Source. For example, you can have an `HTTP` Source and custom `Get` and `Post` Decoders that repeatedly make `GET` and `POST` requests. The user can use the HTTP Source with either one of the Get or Post decoders and potentially also use the `POST` Decoder with other Sources.

Reviewer: This is confusing.
The `Decoder` wouldn't make `GET` or `POST` requests. The `Decoder` would process the response that the `Source` got from one of those requests.

Generally, defining different `Source` and `Decoder` implementations should only be done where their operation is orthogonal, such as the File source, or an HTTP fetching source, where they get back bytes that must then be converted to a struct by *something*.

I'd avoid discussing "writing a `Decoder` for your `Source`" since that implies non-orthogonality between the implementations, in which case the decoding should probably be part of the `Source` itself. (or can be another implementation-specific struct/parameter)

### Putting it all together

> The `dials.Config` function first makes a deep copy of the configuration struct and makes each field a pointer (even the fields in nested structs) with special handling for structs that implement [`encoding.TextUnmarshaler`](https://golang.org/pkg/encoding/#TextUnmarshaler). Then it calls the `Value` method on each Source and stores the returned value. The final step is to to compose the final config struct by overlaying the values from all the different Sources and accounting for the precedence order. Since the fields are pointers, we can directly assign pointers while overlaying. Overlay even has safety checks for deduplicating maps sharing a backing pointer and for structs with self-referential pointers. So when you write your own Source, you just have to pass the Source in to the `dials.Config` function and Dials will take care of deep copying and pointerifying the struct and composing the final struct with overlay.

Reviewer: I feel like the list of step should be a numbered list so the text isn't quite so dense.

	// the dials tag can be used as an alias so when the name in the config file
	// changes, the code doesn't have to change.
	Val2 int `dials:"val_2"`
	// Dials will register a flag with the name matching the dials tag.
	// Without any struct tags, dials will decode the go camel case field name
	// and encode it using lower-kebab-case and use the encoded name as the flag
	// name (ex: val-3). To specify a different flag name, use the `dialsflag`
	// tag. Now, Dials will register a flag with "some-val" name instead.
	// The `dialsdesc` tag is used to provide help message for the flag.
	Val3 bool `dialsflag:"some-val" dialsdesc:"enable auth"`
	// Path holds the value of the path to the config file. Dials follows the

> 	// Go convention for environment variables and will look for the dials tag

Reviewer: Let's make this the `*nix convention for environment variables`. (Screaming snake case is a *nix thing, not a Go thing.)

(with the exception of Java, but that's just wrong)

	// or field name in all caps when struct tags aren't specified. Without any
	// struct tags, it would lookup the PATH environment variable. To specify a

> 	// different env variable, use the `dials_env` tag. Now Dials will lookup

Reviewer: nit: `dialsenv` now.

	// struct doesn't specify a config file to read. We would recommend to use
	// dials.Config directly (shown in the next example) instead of the ez
	// package if you just want to use environment variables and flags without a
	// file source
	return c.Path, true
}

func main() {
	c := &Config{}

	ctx, cancel := context.WithCancel(context.Background())
	defer cancel()

	// The following function will populate the config struct by reading the
	// config files, environment variables, and command line flags (order matches
	// the function name) with increasing precedence. In other words, the flag
	// source (last) would overwrite the YAML source (first) were they both to
	// attempt to set the same struct field. There are several options that can be
	// passed in as well to indicate whether the file will be watched and updates
	// to the file should update the config struct and if the flags name component
	// separation should use different encoding than lower-kebab-case
	d, dialsErr := ez.YAMLConfigEnvFlag(ctx, c, ez.WithWatchingConfigFile(true))
	if dialsErr != nil {
		// error handling
	}

	// Fill will make a deep copy of the struct and populate it with values from

> 	// config file, environment variables, and command line flags. Can

Reviewer: This is a little confusing.
How about: `Fill will deep copy the fully-stacked configuration into its argument. The stacked configuration is populated from the config file, environment variables and commandline flags.`

Also, let's start the next sentence as `One can` rather than `Can`, so there's a subject.

		return
	}

	// Use a non-background context with the WatchingSource to prevent goroutine leaks
	ctx, cancel := context.WithCancel(context.Background())
	defer cancel()

	// additional sources can be passed along with the watching file source and the
	// precedence order will still be dictated by the order in which the sources are
	// defined in the Config function.
	d, err := dials.Config(ctx, config, watchingFileSource)
	if err != nil {
		// error handling
	}

	conf := d.View().(*Config)

> 	// you can get notified whenever the config changes through the channel in

Reviewer: nit: `through the channel returned by the Events method`

### Decoder
Decoders are modular allowing users to mix and match Decoders and Sources. Dials currently supports Decoders that decode different data formats (JSON, YAML, and TOML) and insert the values into the appropriate fields in the config struct. Decoders can be expanded from that use case and users can write their own Decoders to perform the tasks they like (more info in the section below). Decoder is called when the supported Source calls the `Decode` method to unmarshal the data into the config struct and returns the populated struct. There are two sources that the Decoders can be used with: files (including watched files) and `static.StringSource`. Please note that the Decoder interface is likely to change in the near future.

### Write your own Source and Decoder

> If you wish to define your own source, implement the `Source` interface and pass the source to the `dials.Config` function. If you want the Source to interact with a Decoder, call `Decode` in the `Value` method of the Source. Since Decoders are modular, keep the logic of Decoder encapsulated and separate from the Source. `Source` and `Decoder` implementations should be orthogonal and `Decoder`s should not be `Source` specific. For example, you can have an `HTTP` or `File` Source that can interact with the `JSON` decoder to unmarshal the data to a struct.

Reviewer: nit: It would be nice to add a couple line-breaks in here and maybe split this into a couple paragraphs.
